### DISCUSSION TITLE
Client-side Playground is now available!

### DISCUSSION THREAD

**Discussion by endel** [colyseus/colyseus#585] [Show and tell]
https://github.com/colyseus/colyseus/assets/130494/ebd90a9a-784b-44a1-8337-85e6a20cc8e8

## How to test it

The playground tool is installed by default on new projects created via `npm create colyseus-app`.

To manually install on an existing project, please do:

```bash
npm install --save @colyseus/playground
```

Then, import and bind it to your express app instance:

```typescript
// import the route
import { playground } from "@colyseus/playground";

// bind it as an express middleware
app.use("/playground", playground);
```

## Your feedback is welcome!

Is there a feature you'd like to see, or an improvement that would make your life easier inspecting/debugging your room implementation?

**Comment by BugSlayer87** [comment]
Awesome to found this Client-side Playground is now available.
Is there any requirement for using this ?
Because in my colyseus version 0.14 its error

**Reply by endel** [reply]
Hi @BugSlayer87, it only works on version 0.15

**Comment by pastean** [comment]
I was under the impression that this would also show existing rooms, or is this designed with another goal in mind?

**Reply by avatus** [reply]
You can use https://github.com/colyseus/colyseus-monitor to inspect existing rooms, including state. Playground (from what I understand) is more of a way to test interactions from a client <-> room

**Reply by vinkent420** [reply]
Thank @avatus , you helped me a lots. I saw state tab now in monitor. But don't know how to keep that state when user (player) join room again, ex. last position in map to render exactly. 

And I tried to deploy Colyseus server on GCP K8s, when I deploy again, I see the room was disposing (delete in my MongoDB) -> so user will createNewRoom by RoomName, don't know if we have ways to solve this.

Many thank @avatus @endel so so much, so appreciate!

**Comment by Yashko** [comment]
Hello! I'm playing around with Colyseus and I like it so far. Is there a way to share state only partially with a client? Let's say I have Room with locations that needed to be actually reached before client would be able to see what items in there

**Reply by Yashko** [reply]
@endel

**Reply by endel** [reply]
Hi @Yashko, thanks for your interest - right now partial state sync is only supported via the `@filter()`/`@filterChildren()`, which has some limitations and I do not recommend using for anything other than turn-based games as it consumes too much CPU. 

A new "data ownership" model is on the roadmap as part of improvements on the state sync structures - though no progress has been made there yet https://github.com/colyseus/colyseus/wiki/Public-Roadmap#colyseus-framework

**Comment by beiryu** [comment]
Hi @endel , 

Recently, I have been using [colyseus/colyseus](https://github.com/colyseus/colyseus) as a game server and [colyseus/playground](https://github.com/colyseus/playground) to test messages between client-room-server.

Displaying schemas is a necessary feature to allow clients (game clients) to know which properties they will use in the schema to make requests and receive corresponding responses.

I want to use my code to contribute to the [colyseus/playground](https://github.com/colyseus/playground) repository. For those who might want to use a similar feature in the future.

Currently, I have only created a pull request [here](https://github.com/colyseus/playground/pull/6), you can review the code and the attached image.

If possible, please let me know the process for contributing to the [colyseus/playground](https://github.com/colyseus/playground) repository, including code conventions, etc. I will follow them.

Thanks,

![Screenshot_13](https://github.com/user-attachments/assets/f3a23209-8935-47d1-a46c-6b82f560da19)

**Comment by huseyindeniz** [comment]
hi @endel , thanks for the playground repo, and quick feedback for the last version:

I think typescript definition is missing in the latest version (0.15.16). There is no index.d.ts file under the build dir anymore. I also can't find any types package for this playground package. VS Code is suggesting to install npm i --save-dev @types/colyseus__playground but obviously that package is not exist. 

(Issues tab is not active in the playground repo btw, so I can't report it there.)

![image](https://github.com/user-attachments/assets/1a7a4fc1-406f-47d4-8784-1407cf8cbfec)

![image](https://github.com/user-attachments/assets/a56cb337-c47c-468d-902a-2aa9bc906ec3)

![image](https://github.com/user-attachments/assets/5a20ea05-433e-4169-b4fb-57e8d1e63c81)

**Reply by endel** [reply]
Hi @huseyindeniz, thank you for the detailed report. I just published version `0.15.17` fixing this. _(Some of the repo's Issues tab have been disabled so we can use the main "colyseus" repo for all of them...)_
